server/src/main/java/org/elasticsearch/index/mapper/Mapper.java

```diff
@@ -105,4 +109,23 @@ public String toString() {
     public static String internFieldName(String fieldName) {
         return fieldNameStringDeduplicator.deduplicate(fieldName);
     }
+
+    private static final Map<FieldType, FieldType> fieldTypeDeduplicator = new ConcurrentHashMap<>();
+
+    /**
+     * Freezes the given {@link FieldType} instances and tries to deduplicate it as long as the field does not return a non-empty value for
+     * {@link FieldType#getAttributes()}.
+     *
+     * @param fieldType field type to deduplicate
+     * @return deduplicated field type
+     */
+    public static FieldType freezeAndDeduplicateFieldType(FieldType fieldType) {
+        fieldType.freeze();
+        var attributes = fieldType.getAttributes();
+        if ((attributes != null && attributes.isEmpty() == false) || fieldType.getClass() != FieldType.class) {
+            // don't deduplicate subclasses or types with non-empty attribute maps to avoid memory leaks
+            return fieldType;
+        }
+        return fieldTypeDeduplicator.computeIfAbsent(fieldType, Function.identity());
+    }
```

Comment:
I think this is OK if there is a static upper limit on the customization. But for `dense_vector`, this is not the case, each different dimension value would be a different `FieldType`. Somebody thinking they are doing the correct thing and use this method with `dense_vector` will cause a memory leak.

Reply:
Don't we have an upper bound of 1024 for the vector dimension and limits of 8 and 16 on the other dimension `int` parameters? There's not that many possible combinations here (as in a couple thousand at best?) or am I missing something?

Reply:
`2048 * 3 * 2` are the number of combinations for vector focused field types.

This will increase as we will bump our maximum dense_vector dimension in the future as well.